src/main/scala/io/hydrosphere/serving/manager/domain/application/migrations/ApplicationSignatureMigrationTool.scala

```diff
@@ -0,0 +1,147 @@
+import cats.implicits._
+import doobie.util.transactor.Transactor
+import io.hydrosphere.serving.manager.domain.application.{
+  Application,
+  ApplicationKafkaStream,
+  ApplicationRepository,
+  ApplicationServable,
+  GraphComposer
+}
+import io.hydrosphere.serving.manager.domain.model_version.{ModelVersion, ModelVersionRepository}
+import io.hydrosphere.serving.manager.infrastructure.db.repository.{
+  DBApplicationRepository,
+  DBGraph
+}
+import doobie.implicits._
+import io.hydrosphere.serving.manager.domain.DomainError
+import io.hydrosphere.serving.manager.infrastructure.db.repository.DBApplicationRepository.ApplicationRow
+import io.circe.parser._
+import io.hydrosphere.serving.manager.domain.contract.Signature
+import io.hydrosphere.serving.manager.domain.deploy_config.{
+  DeploymentConfiguration,
+  DeploymentConfigurationRepository
+}
+import io.hydrosphere.serving.manager.domain.servable.{Servable, ServableRepository}
+import org.apache.logging.log4j.scala.Logging
+
+trait ApplicationSignatureMigrationTool[F[_]] {
+  def getAndRecover(): F[Unit]
+}
+
+object ApplicationSignatureMigrationTool extends Logging {
+
+  def default[F[_]](
+      appsRepo: ApplicationRepository[F],
+      modelVersionRepo: ModelVersionRepository[F],
+      servableRepo: ServableRepository[F],
+      deploymentRepo: DeploymentConfigurationRepository[F]
+  )(implicit F: Bracket[F, Throwable], tx: Transactor[F]) =
+    new ApplicationSignatureMigrationTool[F] {
+      override def getAndRecover(): F[Unit] = {
+        logger.info("Start recovering process...")
+        for {
+          apps <- DBApplicationRepository.allQ.to[List].transact(tx)
+          appsNeedsToUpdate = apps.filter(ar => decode[Signature](ar.application_contract).isLeft)
+          result <- appsNeedsToUpdate.traverse { appRow =>
+            for {
+              graph <- F.fromEither(decode[DBGraph](appRow.execution_graph))
+              versions <-
+                appRow.used_model_versions
+                  .traverse(modelVersionRepo.get)
+                  .map {
+                    _.collect {
+                      case Some(x: ModelVersion.Internal) => x.id -> x
+                    }.toMap
+                  }
+              servables <-
+                appRow.used_servables
+                  .traverse(servableRepo.get)
+                  .map(list =>
+                    list.collect {
+                      case Some(servable) => servable.fullName -> servable
+                    }.toMap
+                  )
+              depNames = graph.stages.flatMap(_.variants).toList.flatMap(_.requiredDeployConfig)
+              deployments <-
+                depNames
+                  .traverse(deploymentRepo.get)
+                  .map(list =>
+                    list.collect {
+                      case Some(dc) => dc.name -> dc
+                    }.toMap
+                  )
+              _ <- restoreApp(appRow, versions, servables, deployments) match {
```

Comment:
❌  `restoreApp` is an effectful function, right? So, if I put `restoreApp[F](...)` how is it possible to pattern match against `Left` and `Right`?

Reply:
Made it as function which returns Either[DomainError, Application]